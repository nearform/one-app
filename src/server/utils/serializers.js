import { Map as iMap } from 'immutable';

import transit from '../../universal/utils/transit';

// eslint-disable-next-line import/prefer-default-export
export function serializeClientInitialState(clientInitialState) {
  // try to build the full state, this _might_ fail (ex: 'Error serializing unrecognized object')
  try {
    return transit.toJSON(clientInitialState);
  } catch (err) {
    console.error('encountered an error serializing full client initial state', err);

    // clear out an internal cache that corrupts the serialization generated on the next call
    // TODO: understand transit-js and transit-immutable-js internals to properly fix the bug
    // for now stop the bleeding
    transit.toJSON('clear out an internal cache');
  }

  // can't send all the work we've done to build the state, but we can still give the app what it
  // needs to start in the browser
  // this _shouldn't_ ever throw, but just in case...
  try {
    return transit.toJSON(
      iMap({
        config: clientInitialState.get('config'),
        holocron: clientInitialState.get('holocron'),
      })
    );
  } catch (err) {
    transit.toJSON('clear out an internal cache, again');
    // something is really wrong
    console.error('unable to build the most basic initial state for a client to startup', err);
    throw err;
  }
}
