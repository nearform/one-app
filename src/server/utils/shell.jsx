/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import ReactDOM from 'react-dom/server';

const templates = new Map();

export const SSR_SHELL_KEY = 'default';
export const templateAttrsPattern = /data-attrs="([{}A-Z_]*)"(>)/g;
export const templatePattern = /({{{[A-Z_]{1,}}}})/g;
export const templateTags = {
  doctype: '{{{DOC_TYPE}}}\n',
  appHtml: '{{{APP_HTML}}}',
  lang: '{{{LANG}}}',
  htmlAttrs: '{{{HTML_ATTRS}}}',
  headAttrs: '{{{HEAD_ATTRS}}}',
  bodyAttrs: '{{{BODY_ATTRS}}}',
  rootAttrs: '{{{ROOT_ATTRS}}}',
  head: '{{{HEAD}}}',
  title: '{{{DOCUMENT_TITLE}}}',
  base: '{{{BASE_TAG}}}',
  meta: '{{{META}}}',
  links: '{{{LINKS}}}',
  manifest: '{{{MANIFEST}}}',
  styles: '{{{STYLE_TAGS}}}',
  initialState: '{{{INITIAL_STATE}}}',
  scripts: '{{{SCRIPT_TAGS}}}',
};
export const tagDefaults = {
  doctype: () => '<!DOCTYPE html>',
  lang: () => 'en_US',
  rootAttrs: () => 'id="root"',
};

export function getTemplate(type = SSR_SHELL_KEY) {
  return templates.get(type);
}

export function setTemplate(type, template, tags = []) {
  templates.set(type, { template, tags });
  return getTemplate(type);
}

export function setTagDefault(tag, fn) {
  tagDefaults[tag] = fn;
  return tagDefaults;
}

export function renderToStaticMarkup(root, doctype = '') {
  return [doctype, ReactDOM.renderToStaticMarkup(root)].join('');
}

export function renderTemplate(type) {
  const tags = Object.keys(templateTags);
  let template = '';
  const templateType = type || SSR_SHELL_KEY;

  switch (templateType) {
    default:
    case SSR_SHELL_KEY: {
      template = renderToStaticMarkup(
        // eslint-disable-next-line jsx-a11y/html-has-lang
        <html data-attrs={templateTags.htmlAttrs}>
          <head data-attrs={templateTags.headAttrs}>
            {templateTags.head}
            {templateTags.title}
            {templateTags.base}
            {templateTags.meta}
            {templateTags.links}
            {templateTags.manifest}
            {templateTags.styles}
          </head>
          <body data-attrs={templateTags.bodyAttrs}>
            <div data-attrs={templateTags.rootAttrs}>
              {templateTags.appHtml}
            </div>
            {templateTags.initialState}
            {templateTags.scripts}
          </body>
        </html>,
        templateTags.doctype
      ).replace(templateAttrsPattern, '$1$2');
      break;
    }
  }

  if (!templates.has(templateType)) return setTemplate(templateType, template, tags);

  return { template, tags };
}

export function renderShell(replacements = {}, templateKey = SSR_SHELL_KEY, useDefaults = true) {
  const template = templates.get(templateKey) || renderTemplate(templateKey);

  const html = template.tags
    .map((key) => {
      const tag = templateTags[key];
      let value = replacements[key];
      if (!value && value !== null && useDefaults) value = tagDefaults[key] ? tagDefaults[key]() : '';
      if (typeof value === 'string') value = value.trim();
      else value = '';
      return [key, tag, value];
    })
    .reduce(
      // eslint-disable-next-line complexity
      (templateString, [key, tag, value]) => {
        const nextValue = value ? `${templateString.replace(tag, value).trim()}` : templateString;

        if (!value) {
          switch (key) {
            default:
              break;
            case 'appHtml': return nextValue;
            case 'rootAttrs':
            case 'bodyAttrs':
            case 'headAttrs':
            case 'htmlAttrs':
              return nextValue.replace(` ${tag}`, '').replace(/(?:\s*)data-attrs=".*"/, '');
          }
        } else {
          switch (key) {
            default:
              break;
            case 'rootAttrs':
            case 'bodyAttrs':
            case 'headAttrs':
            case 'htmlAttrs':
              return nextValue.replace(/(?:\s*)data-attrs=".*"/, '');
            case 'appHtml':
              return `\n${nextValue}\n`;
          }
        }

        return [nextValue, value && '\n'].join('');
      },
      template.template
    )
    .replace(templatePattern, '');

  return html;
}
