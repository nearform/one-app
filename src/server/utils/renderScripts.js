/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { Set as iSet } from 'immutable';
import uuid from 'uuid/v4';

import jsonStringifyForScript from './jsonStringifyForScript';
import getI18nFileFromState from './getI18nFileFromState';
import readJsonFile from './readJsonFile';
import { getClientModuleMapCache } from './clientModuleMapCache';
import { serializeClientInitialState } from './serializers';

const integrityManifest = readJsonFile('../../../bundle.integrity.manifest.json');

export function renderScriptTag(src, attrs = [], body = '') {
  if (src) attrs.unshift(`src="${src}"`);
  return `<script${[
    attrs.length > 0 ? ' ' : '',
    attrs.join(' ').trim(),
  ].join('')}>${body ? `\n${body}\n` : ''}</script>`;
}

export function renderInitialStateScriptTag({
  id = 'initial-state',
  isStatic = false,
  scriptNonce,
  publicPath,
  bundle,
  moduleMap: __CLIENT_HOLOCRON_MODULE_MAP__,
  initialState: __INITIAL_STATE__,
}) {
  const initialStateMap = {
    __render_mode__: `'${isStatic ? 'render' : 'hydrate'}'`,
    __webpack_public_path__: publicPath,
    __holocron_module_bundle_type__: `'${bundle}'`,
    __CLIENT_HOLOCRON_MODULE_MAP__,
    __INITIAL_STATE__,
  };

  return renderScriptTag(
    null,
    [`id="${id}"`, scriptNonce ? `nonce="${scriptNonce}"` : ''],
    Object.keys(initialStateMap)
      .map((key) => `\twindow.${key} = ${initialStateMap[key]};`)
      .join('\n')
  );
}

export function getModules(modules, moduleMap, rootModuleName) {
  // Sorting to ensure that the rootModule is the first script to load,
  // this is required to correctly provide external dependencies.
  return modules
    .sort((currentModule, nextModule) => {
      if (currentModule === rootModuleName) {
        return -1;
      }
      if (nextModule === rootModuleName) {
        return 1;
      }
      return 0;
    })
    .map((moduleName) => [moduleMap.modules[moduleName], moduleName])
    .toJS();
}


function getChunkAssets(assetsByChunkName) {
  // console.log(assetsByChunkName);
  return (
    Object.entries(assetsByChunkName)
      // i18n is different per request, app needs to be the last chunk loaded
      .filter(
        ([chunkName]) => chunkName !== 'app'
          && /^(i18n)/g.test(chunkName) === false
      )
      .map(([, assets]) => (typeof assets === 'string' ? assets : assets[0]))
  );
}

export const modernBrowserChunkAssets = getChunkAssets(
  readJsonFile('../../../.build-meta.json').modernBrowserChunkAssets
);
export const legacyBrowserChunkAssets = getChunkAssets(
  readJsonFile('../../../.build-meta.json').legacyBrowserChunkAssets
).map((chunkAsset) => `legacy/${chunkAsset}`);

export function renderChunkAssetScripts({
  legacy,
  publicPath,
  crossOrigin,
  integrity,
}) {
  const chunkAssets = legacy ? legacyBrowserChunkAssets : modernBrowserChunkAssets;
  return chunkAssets
    .map((chunkAsset) => renderScriptTag(`${publicPath}/${chunkAsset}`, [
      integrity ? `integrity="${integrityManifest[chunkAsset]}"` : '',
      crossOrigin ? `crossorigin="${crossOrigin}"` : '',
    ]))
    .join('\n');
}

export function renderI18nScript({
  clientInitialState, publicPath, crossOrigin,
}) {
  const i18nFile = getI18nFileFromState(clientInitialState);
  if (!i18nFile) {
    return '';
  }

  return renderScriptTag(`${publicPath}/${i18nFile}`, [
    crossOrigin ? `crossorigin="${crossOrigin}"` : '',
  ]);
}

export function renderModuleScripts({
  clientInitialState,
  rootModuleName = clientInitialState.getIn(['config', 'rootModuleName'], ''),
  modules = clientInitialState.getIn(['holocron', 'loaded'], iSet()),
  moduleMap,
  bundle,
  devMode,
  crossOrigin,
}) {
  return getModules(modules, moduleMap[bundle], rootModuleName)
    .map(
      ([
        {
          [bundle]: { integrity, url: src },
        },
      ]) => {
        const { key } = moduleMap[bundle];
        return renderScriptTag(devMode ? src : `${src}?key=${key}`, [
          crossOrigin ? `crossorigin="${crossOrigin}"` : '',
          devMode ? '' : `integrity="${integrity}"`,
        ]);
      }
    )
    .join('\n');
}

// eslint-disable-next-line complexity
export function renderScripts({
  devMode = false,
  legacy = false,
  isStatic = false,
  crossOrigin = 'anonymous',
  publicPath = '/_/static',
  scriptNonce = uuid(),
  clientInitialState = null,
  rootModuleName = clientInitialState && clientInitialState.getIn(['config', 'rootModuleName'], 'root'),
  clientModuleMapCache = getClientModuleMapCache(),
} = {}) {
  if (!clientInitialState) throw new Error('clientInitialState is required to render scripts');
  // if (!rootModuleName) throw new Error('rootModuleName is required to render scripts');

  const bundle = legacy ? 'legacyBrowser' : 'browser';
  const bundlePrefixForBrowser = legacy ? `${publicPath}/legacy` : publicPath;

  const renderedState = renderInitialStateScriptTag({
    isStatic,
    scriptNonce,
    bundle,
    publicPath: publicPath ? jsonStringifyForScript(publicPath.endsWith('/') ? publicPath : `${publicPath}/`) : '',
    moduleMap: clientModuleMapCache ? jsonStringifyForScript(clientModuleMapCache[bundle]) : '{}',
    initialState: jsonStringifyForScript(serializeClientInitialState(clientInitialState)),
  });
  const assets = publicPath ? renderChunkAssetScripts({
    legacy,
    publicPath,
    crossOrigin,
    integrity: !devMode,
  }) : '';
  const i18n = publicPath ? renderI18nScript({
    clientInitialState,
    crossOrigin,
    publicPath: bundlePrefixForBrowser,
    integrity: !devMode,
  }) : '';
  const moduleScripts = clientModuleMapCache ? renderModuleScripts({
    clientInitialState,
    moduleMap: clientModuleMapCache,
    devMode,
    bundle,
    crossOrigin,
    rootModuleName,
  }) : '';
  const appScript = publicPath ? renderScriptTag(`${bundlePrefixForBrowser}/app.js`, [
    !devMode ? `integrity="${integrityManifest[legacy ? 'legacy/app.js' : 'app.js']}"` : '',
    crossOrigin ? `crossorigin="${crossOrigin}"` : '',
  ]) : '';

  return [renderedState, assets, i18n, moduleScripts, appScript].filter((str) => !!str).join('\n');
}
