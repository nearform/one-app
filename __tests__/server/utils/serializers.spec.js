import { fromJS, Set as iSet } from 'immutable';

import { serializeClientInitialState } from '../../../src/server/utils/serializers';
import transit from '../../../src/universal/utils/transit';

jest.mock('../../../src/universal/utils/transit', () => ({
  toJSON: jest.fn((value) => JSON.stringify(value)),
}));

describe('serializeClientInitialState', () => {
  const rootModuleName = 'test-root';
  const throwError = () => {
    throw new Error('failed to serialize');
  };

  beforeAll(() => {
    jest.spyOn(console, 'error').mockImplementation(() => null);
  });

  beforeEach(() => {
    transit.toJSON.mockClear();
    console.error.mockClear();
  });

  it('should serialize state', () => {
    expect.assertions(4);

    const initialState = fromJS({
      holocron: {
        loaded: iSet(),
      },
    });
    const str = serializeClientInitialState(initialState);
    expect(str).toContain('holocron');
    expect(str).toContain('loaded');
    expect(str).toMatchSnapshot();
    expect(transit.toJSON).toHaveBeenCalledTimes(1);
  });

  it('should fail to serialize state and fallback to defaults', () => {
    expect.assertions(7);

    const initialState = fromJS({
      holocron: {
        loaded: iSet(),
      },
      config: {
        rootModuleName,
      },
    });
    transit.toJSON.mockImplementationOnce(throwError);
    const str = serializeClientInitialState(initialState);
    expect(str).toContain('config');
    expect(str).toContain('rootModuleName');
    expect(str).toContain('holocron');
    expect(str).toContain('loaded');
    expect(str).toMatchSnapshot();
    expect(transit.toJSON).toHaveBeenCalledTimes(3);
    expect(console.error).toHaveBeenCalledTimes(1);
  });

  it('should throw if fallback state fails to serialize', () => {
    const initialState = fromJS({
      holocron: {
        loaded: iSet(),
      },
      config: {
        rootModuleName,
      },
    });
    transit.toJSON
      .mockImplementationOnce(throwError)
      .mockImplementationOnce(() => 'clear out internal state')
      .mockImplementationOnce(throwError);
    expect(() => serializeClientInitialState(initialState)).toThrow();
    expect(transit.toJSON).toHaveBeenCalledTimes(4);
    expect(console.error).toHaveBeenCalledTimes(2);
  });
});
