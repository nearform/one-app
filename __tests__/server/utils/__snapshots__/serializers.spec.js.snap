// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializeClientInitialState should fail to serialize state and fallback to defaults 1`] = `"{\\"config\\":{\\"rootModuleName\\":\\"test-root\\"},\\"holocron\\":{\\"loaded\\":[]}}"`;

exports[`serializeClientInitialState should serialize state 1`] = `"{\\"holocron\\":{\\"loaded\\":[]}}"`;
